File: 백준/Gold/7569. 토마토/토마토.py
from collections import deque
import sys
input = sys.stdin.readline

M, N, H = map(int, input().split())

box = [[list(map(int, input().split())) for n in range(N)] for h in range(H)]

Q = deque()

zero = True

for h in range(H):
    for n in range(N):
        for m in range(M):
            if box[h][n][m] == 1:
                Q.append([h, n, m])
            elif box[h][n][m] == 0 and zero == True:  # 토마토가 없는 경우
                zero = False

if zero == True:
    print(0)
    quit()

while Q:

    v = Q.popleft()

    for d in [[0, -1, 0], [0, 1, 0], [0, 0, -1], [0, 0, 1], [-1, 0, 0], [1, 0, 0]]:
        nh = v[0] + d[0]
        ni = v[1] + d[1]
        nj = v[2] + d[2]

        if 0 <= ni < N and 0 <= nj < M and 0 <= nh < H:
            if box[nh][ni][nj] == 0:
                box[nh][ni][nj] = box[v[0]][v[1]][v[2]] + 1
                Q.append([nh, ni, nj])

res = 0
for h in range(H):
    for n in range(N):
        for m in range(M):
            if box[h][n][m] > res:
                res = box[h][n][m]
            elif box[h][n][m] == 0:
                print(-1)
                quit()

print(res-1)

File: 백준/Gold/7569. 토마토/토마토.js
class Queue {
    constructor() {
        this.items = [];
        this.frontIndex = 0;
        this.backIndex = 0;
    }

    enqueue(item) {
        this.items[this.backIndex] = item;
        this.backIndex++;
    }

    dequeue() {
        if (this.isEmpty()) return null;
        const item = this.items[this.frontIndex];
        this.frontIndex++;
        return item;
    }

    isEmpty() {
        return this.frontIndex === this.backIndex;
    }

    size() {
        return this.backIndex - this.frontIndex;
    }
}

const getTomato = (N, M, H, boxes) => {
    let queue = new Queue();
    for (let h = 0; h < H; h++) {
        for (let n = 0; n < N; n++) {
            for (let m = 0; m < M; m++) {
                if (boxes[h][n][m] === 1) {
                    queue.enqueue([h, n, m]);
                }
            }
        }
    }
    return queue;
}

const solve = (N, M, H, boxes) => {
    let queue = getTomato(N, M, H, boxes);
    const directions = [[1, 0, 0], [-1, 0, 0], [0, 1, 0], [0, -1, 0], [0, 0, 1], [0, 0, -1]];

    let days = -1;

    while (!queue.isEmpty()) {
        let size = queue.size();
        while (size > 0) {
            const [h, n, m] = queue.dequeue();
            for (const [dh, dn, dm] of directions) {
                const nh = h + dh, nn = n + dn, nm = m + dm;

                if (0 <= nh && nh < H && 0 <= nn && nn < N && 0 <= nm && nm < M) {
                    if (boxes[nh][nn][nm] === 0) {
                        boxes[nh][nn][nm] = 1;
                        queue.enqueue([nh, nn, nm]);
                    }
                }
            }
            size--;
        }
        days++;
    }

    for (let h = 0; h < H; h++) {
        for (let n = 0; n < N; n++) {
            for (let m = 0; m < M; m++) {
                if (boxes[h][n][m] === 0) {
                    return -1;
                }
            }
        }
    }

    return days;
}

const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let input = [];
rl.on('line', line => {
    input.push(line);
    if (Number(input[0].split(' ')[1]) * Number(input[0].split(' ')[2]) < input.length - 1) rl.close();
}).on('close', () => {
    const [M, N, H] = input[0].split(' ').map(Number);
    const boxes = [];
    for (let i = 1; i < input.length; i += N) {
        boxes.push(input.slice(i, i + N).map(e => e.split(' ').map(Number)));
    }
    console.log(solve(N, M, H, boxes));
});
